internal/guest/runtime/hcsv2/spec.go

```diff
@@ -152,3 +167,68 @@ func getGroup(spec *oci.Spec, filter func(user.Group) bool) (user.Group, error)
 	}
 	return groups[0], nil
 }
+
+// applyAnnotationsToSpec modifies the spec based on additional information from annotations
+func applyAnnotationsToSpec(ctx context.Context, spec *oci.Spec) error {
+	// Check if we need to override container's /dev/shm
+	if val, ok := spec.Annotations["io.microsoft.container.storage.shm.size-kb"]; ok {
+		sz, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return errors.Wrap(err, "/dev/shm size must be a valid integer")
+		}
+		if sz <= 0 {
+			return errors.Errorf("/dev/shm size must be a positive integer, got: %d", sz)
+		}
+
+		// Use the same options as in upstream https://github.com/containerd/containerd/blob/0def98e462706286e6eaeff4a90be22fda75e761/oci/mounts.go#L49
+		size := fmt.Sprintf("size=%dk", sz)
+		mt := oci.Mount{
+			Destination: "/dev/shm",
+			Type:        "tmpfs",
+			Source:      "shm",
+			Options:     []string{"nosuid", "noexec", "nodev", "mode=1777", size},
+		}
+		spec.Mounts = removeMount("/dev/shm", spec.Mounts)
+		spec.Mounts = append(spec.Mounts, mt)
+	}
+
+	// Check if we need to do any capability/device mappings
+	if spec.Annotations["io.microsoft.virtualmachine.lcow.privileged"] == "true" {
+		log.G(ctx).Debug("'io.microsoft.virtualmachine.lcow.privileged' set for privileged container")
+
+		// Add all host devices
+		hostDevices, err := devices.HostDevices()
+		if err != nil {
+			return err
+		}
+		for _, hostDevice := range hostDevices {
+			addLinuxDeviceToSpec(ctx, hostDevice, spec, false)
```

Comment:
would this not override what you did above when the container is privileged?

Reply:
it looks like `/dev/shm`  is skipped here https://github.com/microsoft/hcsshim/blob/a8f5f611a65940f3f974697105e000497b49bbea/vendor/github.com/opencontainers/runc/libcontainer/devices/device_unix.go#L91-L92, so our override should stay. I also did a quick test to validate.